### PR TITLE
textarea widget

### LINKED ISSUE #44: Textarea widget
**Enhancement** 
Related to #5 

Features:
- Includes styles for disabled, required, valid, and invalid states
- accepts callback functions for:
  - `onBlur`
  - `onChange`
  - `onClick`
  - `onFocus`
  - `onInput`
  - `onKeyDown`
  - `onKeyPress`
  - `onKeyUp`
  - `onMouseDown`
  - `onMouseUp`
  - `onTouchStart`
  - `onTouchEnd`
  - `onTouchCancel`
- value is controlled by the parent component (example included)

### PR DESCRIPTION
**Pull request by smhigley**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Resolves #44, adds textarea widget

### CODE REVIEW

File: src/textarea/Textarea.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	render() {
+		const {
+			columns = null,
+			rows = null,
+			wrapText = null
+		} = this.properties;
+
+		return v('textarea', {
```
Comment (bitpshr) [inline src/textarea/Textarea.ts]: How would a user set a default value on this?

Comment (smhigley) [inline src/textarea/Textarea.ts]: `dojo/widget-core/mixins/FormLabel` is where most of the common form node attributes, including `value`, are added.

File: src/textarea/Textarea.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	onKeyUp?(event: KeyboardEvent): void;
+	onMouseDown?(event: MouseEvent): void;
+	onMouseUp?(event: MouseEvent): void;
+	onTouchStart?(event: TouchEvent): void;
+	onTouchEnd?(event: TouchEvent): void;
+	onTouchCancel?(event: TouchEvent): void;
+}
+
+@theme(css)
+export default class Textarea extends ThemeableMixin(FormLabelMixin(WidgetBase))<TextareaProperties> {
```
Comment (agubler) [inline src/textarea/Textarea.ts]: Can you convert this to use the "base" pattern? Create the mixin and assign to a `const` and use that to extend. If you run `grunt dist` this should actually fail as it is.

File: src/textarea/Textarea.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+			typeof invalid === 'boolean' ? invalid ? css.invalid : css.valid : null
+		];
+
+		return v('textarea', {
+			classes: this.classes(...classes).get(),
+			cols: columns ? columns + '' : null,
+			rows: rows ? rows + '' : null,
+			wrap: wrapText,
+			onblur: onBlur,
+			onchange: onChange,
```
Comment (agubler) [inline src/textarea/Textarea.ts]: Does this work? Don't we need to define these in the widget class and call `this.onChange` for example? If the consumer changed the function passed to the widget then I believe Maquette would throw an error (you're not allowed to change the function for handlers).

File: src/textarea/example/index.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
+import { WidgetBase } from '@dojo/widget-core/WidgetBase';
+import { WidgetProperties } from '@dojo/widget-core/interfaces';
+import { ProjectorMixin } from '@dojo/widget-core/mixins/Projector';
+import { StatefulMixin } from '@dojo/widget-core/mixins/Stateful';
+import { v, w } from '@dojo/widget-core/d';
+import Textarea from '../../textarea/Textarea';
+
+export class App extends StatefulMixin(WidgetBase)<WidgetProperties> {
```
Comment (agubler) [inline src/textarea/example/index.ts]: Same applies for the base pattern.

File: src/textarea/Textarea.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	onKeyUp?(event: KeyboardEvent): void;
+	onMouseDown?(event: MouseEvent): void;
+	onMouseUp?(event: MouseEvent): void;
+	onTouchStart?(event: TouchEvent): void;
+	onTouchEnd?(event: TouchEvent): void;
+	onTouchCancel?(event: TouchEvent): void;
+}
+
+@theme(css)
+export default class Textarea extends ThemeableMixin(WidgetBase)<TextareaProperties> {
```
Comment (bitpshr) [inline src/textarea/Textarea.ts]:
```ts
export const TextareaBase = ThemeableMixin(WidgetBase);

export default class Textarea extends TextareaBase<TextareaProperties> {
```

**Pull request merged**